### ISSUE TITLE
Making “search as you type” accessible

### ISSUE THREAD

**Issue by knowler** [npmx-dev/npmx.dev#1653]
A fast search experience is a core feature of npmx. There are a few ways we achieve this currently:

1. The search field is _autofocused_.
2. The search results appear _as you type_. The app does not use a combobox for this. Instead it navigates to the search results page and updates the results as the user types.
3. Pressing <kbd>Enter</kbd> will navigate to the first result. Arrow keys are used to navigate between results.

This functionality is a departure from conventional search experiences in a few ways. While autofocusing the search field is not uncommon for search engines—in fact, just about every search engine does this—what is unconventional is navigating to the search results page as the user types and our current implementation is at least not sufficient to conform to WCAG 2.2.

## Part 1: advising the user and making the behaviour togglable

[WCAG 2.2 Success Criterion 3.2.2 On Input (Level A)](https://www.w3.org/WAI/WCAG22/quickref/#on-input) states:

> Changing the setting of any user interface component does not automatically cause a change of context unless the user has been advised of the behavior before using the component.

That lays out the minimum of what we need to do: _advise the user of the behaviour beforehand_.

Since, the field is autofocused and/or can be navigated to directly by some assistive technologies (e.g. shortcuts or a rotor in screen readers), we should associate this advisory content using `aria-describedby`. We also do need to ensure this _is visible_ as WCAG is not just guidelines for screen reader content.

Now, I think the content of the advisory should say something like: 

> “Instant search: page will navigate to results as you type.”

To not just meet the minimum, it would be good if we made the _search as you type_ functionality a setting (which can default to _on_), so if we could add to the advisory:

> “Turn this off in settings”

The “settings” bit can be a link. Usually we avoid interactive content inside of `aria-describedby` content, however, the text still makes sense without it.

Further, it would be a burden for these users to navigate to the settings before using the app and so I think we should allow the setting to toggled with a shortcut. Something like <kbd><kbd>Alt</kbd>+<kbd>/</kbd></kbd> could work (I have not checked it for conflicts). So, then we’d add to the advisory:

> “Turn this off in settings or press <kbd><kbd>Alt</kbd>+<kbd>/</kbd></kbd>.”

Then since we’re changing a setting, we need some feedback for success and so we could use a live region to indicate that, yes, the setting has changed (and what it has changed to).

So, the full advisory would be:

> Instant search: page will navigate to results as you type. Turn this off in settings or press <kbd><kbd>Alt</kbd>+<kbd>/</kbd></kbd>.

Ideally, in the DOM order this would be placed _before_ the search box, so a user navigating to it would find it beforehand, however, if we did so, we could probably make an exception to the DOM order matches visual order rule and visually place it under the search box (since we don’t have an autocomplete dropdown, placing below seems ok as there wouldn’t be an issue with it being covered).

### TL;DR

The code should be updated to look something like this while the _search as you type_ behaviour is enabled:

```html
<label for=search>search packages</label>
<p id=instant-search-advisory>Instant search: page will navigate to results as you type. Turn this off in <a href=/settings>settings</a> or press <kbd><kbd>Alt</kbd>+<kbd>/</kbd></kbd>.</p>
<input id=search type=search name=q aria-describedby=instant-search-advisory>
```

Give or take some design work, it would look something like this…

<img width="615" height="340" alt="The home page searchbox featuring the aforementioned advisory below." src="https://github.com/user-attachments/assets/338a5e70-e5e3-4fb0-9c40-2d9254495774" />

## Part 2: improving the results page (work-in-progress)

One challenge with switching to the results page is that the context is changed both in the sense of the URL and the location of the input (i.e. it moves from the `main` landmark to the `banner` landmark). In NVDA at least, this is announced which may or may not be sufficient for users to understand where they are so they can navigate to the results. I’m worried this could be disorienting, but at the same time, it wouldn’t be an uncommon search-engine type page layout. Since there’d be a lot happening at once, it is possible some of these announcements could get clobbered by the announcing of the input in the field. In any case, here’s what I think is important here:

- Helping the user reorient themselves on the search result page.
- Helping the user understand when the results are changing.
- Helping the user understand how to get to the results, especially if there are any shortcuts (note that the up/down arrow shortcuts we do have would be useless for some screen reader users as those keys are used for navigating the page).
- Helping the user understand that <kbd>Enter</kbd> might do something unexpected. Perhaps notifying them that pressing <kbd>Enter</kbd> will choose the first result (and naming it) could be explored.

**Comment by serhalp** [comment]
@knowler do you think it's realistic to aim to address this by the March 3 launch? in any case, is this research in a place where you could write up some issues that folks can start working on?

**Comment by knowler** [comment]
@serhalp I had some feedback from some trusted folks that I’ll incorporate this morning and maybe create some sub-issues for. I do think it should be feasible.

**Comment by knowler** [comment]
#1845 was never linked as a sub-issue, but it was one.

Its implementation was not satisfactory for resolving this issue. The following was my assessment (https://github.com/npmx-dev/npmx.dev/pull/1845#issuecomment-3991019503):

> We’ll need to follow up with a few changes to the wording as it’s too vague.
>
> There are two parts here: the description of the behaviour and the description of how to turn it off.
>
> “Instant search” on its own is vague, but it might be fine if we can direct users to a more clear definition. Though this is the main part that is required by WCAG 2.2 SC 3.2.2.
>
> What I think is almost more important is the description of how to turn it off. The problem with just using an action “turn off” is that this doesn’t make a whole lot of sense for screen reader users. aria-describedby does not convey semantics or even location of the description. So, whatever’s set for it needs to make sense on its own. This is why I had worded it to direct the user where to go (“Turn off in settings”) and gave them an immediate way to shut it off (the hotkey).
>
> We could consider using a tooltip to reduce the visual noise here and then provide a screen reader only description for the control.

So, effectively this issue is blocked by our tooltip solution not being properly accessible (an issue I’ve been meaning to file with some research I did).
